autosklearn/data/feature_validator.py

```diff
@@ -0,0 +1,422 @@
+                checks) and a encoder fitted in the case the data needs encoding
+        """
+        if hasattr(X, "iloc") and not scipy.sparse.issparse(X):
+            X = typing.cast(pd.DataFrame, X)
+            # Treat a column with all instances a NaN as numerical
+            # This will prevent doing encoding to a categorical column made completely
+            # out of nan values -- which will trigger a fail, as encoding is not supported
+            # with nan values.
+            # Columns that are completely made of NaN values are provided to the pipeline
+            # so that later stages decide how to handle them
+            if np.any(pd.isnull(X)):
+                for column in X.columns:
+                    if X[column].isna().all():
+                        X[column] = pd.to_numeric(X[column])
+
+            self.enc_columns, self.feat_type = self._get_columns_to_encode(X)
+
+            if len(self.enc_columns) > 0:
+
+                self.encoder = make_column_transformer(
+                    (preprocessing.OrdinalEncoder(
+                        handle_unknown='use_encoded_value',
+                        unknown_value=-1,
+                    ), self.enc_columns),
+                    remainder="passthrough"
+                )
+
+                # Mypy redefinition
+                assert self.encoder is not None
+                self.encoder.fit(X)
+
+                # The column transformer reoders the feature types - we therefore need to change
+                # it as well
+                def comparator(cmp1: str, cmp2: str) -> int:
+                    if (
+                        cmp1 == 'categorical' and cmp2 == 'categorical'
+                        or cmp1 == 'numerical' and cmp2 == 'numerical'
+                    ):
+                        return 0
+                    elif cmp1 == 'categorical' and cmp2 == 'numerical':
+                        return -1
+                    elif cmp1 == 'numerical' and cmp2 == 'categorical':
+                        return 1
+                    else:
+                        raise ValueError((cmp1, cmp2))
+                self.feat_type = sorted(
+                    self.feat_type,
+                    key=functools.cmp_to_key(comparator)
+                )
+        return self
+
+    def transform(
+        self,
+        X: SUPPORTED_FEAT_TYPES,
+    ) -> np.ndarray:
+        """
+        Validates and fit a categorical encoder (if needed) to the features.
+        The supported data types are List, numpy arrays and pandas DataFrames.
+
+        Parameters
+        ----------
+            X_train: SUPPORTED_FEAT_TYPES
+                A set of features, whose categorical features are going to be
+                transformed
+
+        Return
+        ------
+            np.ndarray:
+                The transformed array
+        """
+        if not self._is_fitted:
+            raise NotFittedError("Cannot call transform on a validator that is not fitted")
+
```

Comment:
I believe there is a `self._check_data(X)` missing here.

Reply:
Thanks, I have added this so we can check for test data in case it is provided not through the fit method